### DISCUSSION TITLE
LazyVim: No useful diagnogstics ever show

### DISCUSSION THREAD

**Discussion by mlp1802** [mrcjkb/rustaceanvim#311] [Q&A]
### Have you read the docs and searched existing issues?

- [ ] I have read [the README's troubleshooting guide](https://github.com/mrcjkb/rustaceanvim/blob/master/README.md#troubleshooting).
- [X] I have searched [existing issues](https://github.com/mrcjkb/rustaceanvim/issues?q=is%3Aopen+is%3Aissue+label%3Abug).
- [X] I have searched [existing discussions](https://github.com/mrcjkb/rustaceanvim/discussions).
- [ ] I have searched the existing issues of other plugins related to this issue (if applicable).
- [X] I have searched the [existing rust-analyzer issues](https://github.com/rust-lang/rust-analyzer/issues?q=is%3Aissue+) (if applicable).
- [ ] I can reproduce the issue with stable Neovim (not a distribution or nightly) and the [minimal config](https://github.com/mrcjkb/rustaceanvim/blob/master/troubleshooting/minimal.lua).

### Neovim version (nvim -v)

10

### Operating system/version

Ubuntu

### Output of :checkhealth rustaceanvim

```console
==============================================================================
rustaceanvim: require("rustaceanvim.health").check()

Checking for Lua dependencies ~
- WARNING dap not installed. Needed for debugging features [mfussenegger/nvim-dap](https://github.com/mfussenegger/nvim-dap)

Checking external dependencies ~
- OK rust-analyzer: found rust-analyzer 0.3.1782-standalone
- OK Cargo: found cargo 1.79.0-nightly (d438c80c4 2024-03-19)
- OK rustc: found rustc 1.79.0-nightly (0ad927c0c 2024-03-21)
- OK debug adapter: found codelldb 

Checking config ~
- OK No errors found in config.

Checking for conflicting plugins ~
- OK No conflicting plugins detected.

Checking for tree-sitter parser ~
- OK tree-sitter parser for Rust detected.
```

### How to reproduce the issue

```console
Open my neovim :)
```

### Expected behaviour

  Diagnogstics should show
![dia](https://github.com/mrcjkb/rustaceanvim/assets/4717892/c1a503b4-fad9-4dc1-af77-3f20c01b6f62)

### Actual behaviour

After upgrading lazyvim no helpful diagnogstics other than missing semicolon etc is shown.
I'm not sure, but I think lazyvim just recently began to use this plugin, as after some digging I found that it completely ignored my custom rust_analyzer lsp setup.
I know have a complete replica up this setup, as described on lazyvims website.
https://www.lazyvim.org/extras/lang/rust
 Everything works, except diagnogstics. I simply get the message "no diagnogstics found".

### The minimal config used to reproduce this issue.

```Lua
local common = require("config.common")
return {
  "mrcjkb/rustaceanvim",
  enabled = true,
  version = "^4", -- Recommended
  ft = { "rust" },
  opts = {
    server = {
      on_attach = function(_, bufnr)
        vim.keymap.set("n", "<M-3>", common.workspace_diagnostics, { noremap = true, silent = true })
        vim.keymap.set("n", "<M-w>", common.document_diagnostics, { noremap = true, silent = true })
        vim.keymap.set("n", "<M-w>", common.document_diagnostics, { noremap = true, silent = true })
        vim.keymap.set("n", "<leader>cR", function()
          vim.cmd.RustLsp("codeAction")
        end, { desc = "Code Action", buffer = bufnr })
        vim.keymap.set("n", "<leader>dr", function()
          vim.cmd.RustLsp("debuggables")
        end, { desc = "Rust debuggables", buffer = bufnr })
      end,
      default_settings = {
        -- rust-analyzer language server configuration
        ["rust-analyzer"] = {
          cargo = {
            allFeatures = true,
            loadOutDirsFromCheck = true,
            runBuildScripts = true,
          },
          -- Add clippy lints for Rust.
          checkOnSave = {
            allFeatures = true,
            command = "check",
            extraArgs = { "--no-deps" },
          },
          procMacro = {
            enable = true,
            ignored = {
              ["async-trait"] = { "async_trait" },
              ["napi-derive"] = { "napi" },
              ["async-recursion"] = { "async_recursion" },
            },
          },
        },
      },
    },
  },
  config = function(_, opts)
    vim.g.rustaceanvim = vim.tbl_deep_extend("keep", vim.g.rustaceanvim or {}, opts or {})
  end,
}
```

**Comment by mrcjkb** [comment]
Hey 👋 

Please don't check the checkboxes without reading them.

Converting to a Q&A discussion, as this is a LazyVim issue, not a rustaceanvim bug.

I don't have the capacity to support all the Neovim distributions out there, so I can only suggest the following:

- Follow the troubleshooting guide starting from a minimal config and plain neovim, as recommended in the bug report form.
- Open a Q&A discussion in the LazyVim repo. People who use LazyVim will be more likely to be able to help you than I am 😀

Edit: Your `checkOnSave.command` looks very fishy by the way:

1. It's the wrong command for clippy
2. You shouldn't need to set anything for `checkOnSave`, as rustaceanvim will automatically detect if you have clippy installed and then set it for you.

**Reply by mlp1802** [reply]
Ok...but I spent like the last 10 hours on the problem, I do think I've read anything there is to read about it. I'm not sure how I should know if it's lazy vim or rustaceanvim? could be one or the other, or a combination of both. I posted there as well.

**Reply by mlp1802** [reply]
Edit: I tried used clippy as command, but that doesnt show errors as well, on top of that it shows a tons of warnings AS ERRORS..I did use check before and it has never been a problem.

**Reply by mrcjkb** [reply]
> I do think I've read anything there is to read about it. 

You checked the box

> - [x] I can reproduce the issue with stable Neovim (not a distribution or nightly) and the minimal config.

but the config you posted is a LazyVim config.

> I'm not sure how I should know if it's lazy vim or rustaceanvim?

That's why the issue form says to try to reproduce it with plain, stable neovim and a minimal config. Distributions like LazyVim come with far too many plugins and customisations for me to be able to reason about what's happening.

**Reply by mlp1802** [reply]
OK got it..I thought perhaps there was a quick workaround, that's why I asked. But I understand that there is not. Best thing is probably to drop lazyvim and start over. But I'm too..lazy.

**Reply by mrcjkb** [reply]
What happens if you remove all of the configs?
Afaik, LazyVim does some funky stuff with merging configs, which can lead to things getting overwritten (depending on the order in which it's applied). That can be a bit of a foot gun.

**Reply by mlp1802** [reply]
What configs specifically?I removed old lsp config, so it now only uses rustace. And it works, everything works..it even underlines errors with red..it just doesnt show the actual error messages anywhere.

**Reply by mrcjkb** [reply]
> it even underlines errors with red..it just doesnt show the actual error messages anywhere.

That sounds like another plugin or module in LazyVim is calling [`vim.diagnistic.config`](https://neovim.io/doc/user/diagnostic.html#vim.diagnostic.config()).

**Reply by mlp1802** [reply]
Yes that could be but it's strange that it works when you type gibberish, as you can see in the picture. It's not only virtual text that doesnt work, also the list of errors in telescope is empty, and I get the message "no dianogstics found".  Anyway it's impossible to fix I think, and right now I'm looking for other editors, until I find a way,  as I'm on a tight deadline. Could of course start with a fresh vim, but I'm scared the problem will still be there.  I'm sure it will somehow work down the road.

**Reply by mrcjkb** [reply]
Based on your screenshot, it looks like you're editing standalone files, which are not part of a Cargo project?

**Reply by mlp1802** [reply]
No they are part of a huge workspace. I posted on lazyvims GitHub as well, no one there knew what was wrong. I ended up reinstalling lazy vim with default plugins, enabling them one by one. Nothing helped. It's crazy but I can't use vim anymore lol, so I'm looking into other editors. Fired up my old spacemacs for the first time in years, still works, and with Rust as well. I just never seen such incomprehensible and confusing documentation as in the emacs world, and half the packages dont seem to be maintained, but still works lol. Anyway I'm sure it will be resolved with time some how.

**Reply by mrcjkb** [reply]
Hmm. I was able to reproduce similar behaviour in a standalone file, but not in a cargo project.

**Reply by ArcusCerebellumus** [reply]
Do you have your project setup as a lib with a bin folder for compiling multiple executables?

Because I do, and I have the same issue, whereas the default style project with main.rs has no problem with diagnostics. And even in my non-orthodox (read it as "jank" if you prefer) setup when I edit anything that's within the scope of my crate (i.e. included somewhere via mod/use) - the diagnostics work just fine.

PS: ...and just like in your case, before I moved to use this plugin (instead of manually setting everything up) and LSP provided from Rustup (instead of one that Mason ships) all of those issues didn't exist for me. :|

**Reply by mrcjkb** [reply]
@ArcusCerebellumus

> before I moved to use this plugin (instead of manually setting everything up) and LSP provided from Rustup (instead of one that Mason ships) all of those issues didn't exist for me.

I don't understand this sentence. Did you use rustup before, and are using mason with this plugin now? Or the other way around?

**Reply by ArcusCerebellumus** [reply]
I used Mason for getting rust_analizer before, just today I moved away from it and got what seems to be the exact same issue as the OP, among others.

What's worse is that restructuring is not really an option at this point :|

**Reply by mrcjkb** [reply]
It's still possible to use mason.nvim's installation with this plugin. You just need to set the `server.cmd` manually (getting the path to the binary from mason-registry) and make sure mason-lspconfig (if you're using it) doesn't set up a conflicting nvim-lspconfig client.

See [`:h rustaceanvim.mason`](https://github.com/mrcjkb/rustaceanvim/blob/master/doc/mason.txt)

You can also try running `cargo clean` in your project. One reason it might have been working with mason in your case, is that when rust-analyzer has been built with a different toolchain than your project, it needs to recompile everything.

**Reply by ArcusCerebellumus** [reply]
Checked your recompilation theory - didn't help. 

But you gave me an idea. From the results of what I did, it became evermore obvious that the compiler KNOWS where everything is - it's the LSP who's clueless. I thought, "Maybe I should help it a little," and since its main function is semantic analysis - I decided to be AS CLEAR as possible. 

Fixed up my bin files by adding explicit `extern` declarations (as opposed to the implicit ones as they were before):

``` diff
+ extern crate blah;
```

Also had to massage my imports a bit, 

``` diff
- use old_name::{
+ use blah::{
```

Also-also I had a naming "conflict" between my package and one of the binaries - fixed that too.

Somewhat of a needless exercise, but now LSP is happy! Throwing error messages like it's no tomorrow :D

**Comment by op8867555** [comment]
Hi,
I ran into a similar problem (#324), I've tried the minimal config setup or downgrading lazy.nvim but still not working :/
Any suggestions to diagnostic this?

**Reply by mrcjkb** [reply]
I would need some more info. E.g. the exact output of `:checkhealth rustaceanvim` and a link to a project that can be used to reproduce the behaviour.

**Reply by op8867555** [reply]
sorry, I'm afraid I can't to do that since it happens in my workplaces' air-gapped environment. I'm trying to create a MRE at home, with the exact same config, but I haven't to manage to reproduce it yet. I guess I can provide version output by check health, Does this help?

**Reply by mrcjkb** [reply]
I'm afraid there's not much I can do without a project to reproduce it with.

Have you read through the messages in the other thread in this topic, for example, [this comment](https://github.com/mrcjkb/rustaceanvim/discussions/311#discussioncomment-8895025)?
Maybe there's something that can help you there?

**Reply by op8867555** [reply]
Thanks, I'll check it once I get to office tomorrow. I'm wondering what are the possibilities that makes rust-tools.nvim working but rustaceanvim not working.

**Reply by mrcjkb** [reply]
> I'm wondering what are the possibilities that makes rust-tools.nvim working but rustaceanvim not working.

For most people, it's a discrepancy between the rust-analyzer version being used. A lot of people use mason.nvim or mason-lspconfig, which I don't recommend, especially for rust (see `:h rustaceanvim.mason`). This plugin defaults to picking up what it finds on the PATH, which is usually a rust-analyzer installation managed by rustup.

There might be some differences in the client capabilities registered by the two plugins, but so far nothing I'm aware of.

**Reply by op8867555** [reply]
Thanks for the explanation, I checked the both capabilities reported by RA but not much difference there.

It turns out the problem is caused by clippy in my case, rust-tools.nvim didn't support clippy and my codebase has some clippy errors. After I turn off enable_clippy, cargo-check diagnostics shows up again. It seems there won't be any cargo-check diagnostics when clippy generating errors.

**Reply by mrcjkb** [reply]
Thanks for the insight. That's good to know.

**Reply by Sotatek-TruongNguyen2** [reply]
I have a similar problem @op8867555 . May i know how to ` turn off enable_clippy` ?

**Reply by mrcjkb** [reply]
See `:h rustaceanvim.config` and `:h rustaceanvim.tools.Opts`.

**Reply by Sotatek-TruongNguyen2** [reply]
It's working for me now. Thanks u @mrcjkb Really appreciate that

**Comment by bryanmaina** [comment]
This just happened to me I would like to add to procedure that checks if everything is ok before playing with any rustaceanvim option:
"You should make sure that the project builds in the same terminal you are running vim/neovim in. And if you are running sessions(tmux for example) try to build in that same session you are running the editor in. If you have an issue building in that session clippy will not work."

In my case, for some reason, I run : `export PKG_CONFIG_PATH=/usr/lib/x86_64-linux-gnu/pkgconfig` to solve some build problems with openssl-sys on WSL and then reopened a tmux session where I run neovim. l but that export was not picked up by the tmux session.

It took me 3 hours to find the PKG_CONFIG_PATH was not set in any tmux session I reopened and that is why clippy was failing and no diagnostic were shown but it was working fine outside of those sessions.
